Simplify multiplier display implementation

peppy opened the pull request:
- [x] Depends on https://github.com/ppy/osu/pull/24799

Split out to minimise LoC changes.


https://github.com/ppy/osu/assets/191335/a872b73f-6d6f-402c-b8a3-8613fea25717

peppy committed:
Combine `ModCounterDisplay` and `DifficultyMultiplierDisplay` into one class

peppy committed:
Tidy up `DifficultyMultiplierDisplay`

peppy committed:
Split base design out of `BeatmapAttributesDisplay`

peppy committed:
Update `DifficultyMultiplierDisplay` to use new shared design

bdach committed:
Merge branch 'master' into mod-select-diff-multiplier-visual-matching

bdach commented on osu.Game/Overlays/Mods/ModFooterInformationDisplay.cs:
```diff
@@ -0,0 +1,153 @@
+        protected FillFlowContainer RightContent { get; private set; } = null!;
+        protected Container Content { get; private set; } = null!;
+
+        public BindableBool Collapsed { get; } = new BindableBool();
```
I don't really super agree with putting the `Collapsed` stuff here, given that the difficulty multiplier display doesn't do anything with it. Was there a particular reason for it? Mind if I move it back to `BeatmapAttributesDisplay`?

bdach replied:
Appears like a straightforward move to me, with no dodgy business as far as I can tell...

For reference:

```diff
diff --git a/osu.Game/Overlays/Mods/BeatmapAttributesDisplay.cs b/osu.Game/Overlays/Mods/BeatmapAttributesDisplay.cs
index 8828e017fd..44c29e313b 100644
--- a/osu.Game/Overlays/Mods/BeatmapAttributesDisplay.cs
+++ b/osu.Game/Overlays/Mods/BeatmapAttributesDisplay.cs
@@ -16,6 +16,7 @@
 using osu.Game.Rulesets.Mods;
 using osuTK;
 using System.Threading;
+using osu.Framework.Input.Events;
 using osu.Game.Configuration;
 
 namespace osu.Game.Overlays.Mods
@@ -39,6 +40,8 @@ public partial class BeatmapAttributesDisplay : ModFooterInformationDisplay
         [Resolved]
         private Bindable<IReadOnlyList<Mod>> mods { get; set; } = null!;
 
+        public BindableBool Collapsed { get; } = new BindableBool(true);
+
         private ModSettingChangeTracker? modSettingChangeTracker;
 
         [Resolved]
@@ -47,10 +50,7 @@ public partial class BeatmapAttributesDisplay : ModFooterInformationDisplay
         private CancellationTokenSource? cancellationSource;
         private IBindable<StarDifficulty?> starDifficulty = null!;
 
-        public BeatmapAttributesDisplay()
-        {
-            Collapsed.Value = true;
-        }
+        private const float transition_duration = 250;
 
         [BackgroundDependencyLoader]
         private void load()
@@ -75,6 +75,7 @@ private void load()
                 }
             });
 
+            RightContent.Alpha = 0;
             RightContent.AddRange(new Drawable[]
             {
                 circleSizeDisplay = new VerticalAttributeDisplay("CS") { Shear = new Vector2(-shear, 0), },
@@ -98,6 +99,43 @@ protected override void LoadComplete()
             }, true);
 
             BeatmapInfo.BindValueChanged(_ => updateValues(), true);
+
+            Collapsed.BindValueChanged(_ =>
+            {
+                // Only start autosize animations on first collapse toggle. This avoids an ugly initial presentation.
+                startAnimating();
+                updateCollapsedState();
+            });
+
+            updateCollapsedState();
+        }
+
+        protected override bool OnHover(HoverEvent e)
+        {
+            startAnimating();
+            updateCollapsedState();
+            return true;
+        }
+
+        protected override void OnHoverLost(HoverLostEvent e)
+        {
+            updateCollapsedState();
+            base.OnHoverLost(e);
+        }
+
+        protected override bool OnMouseDown(MouseDownEvent e) => true;
+
+        protected override bool OnClick(ClickEvent e) => true;
+
+        private void startAnimating()
+        {
+            Content.AutoSizeEasing = Easing.OutQuint;
+            Content.AutoSizeDuration = transition_duration;
+        }
+
+        private void updateCollapsedState()
+        {
+            RightContent.FadeTo(Collapsed.Value && !IsHovered ? 0 : 1, transition_duration, Easing.OutQuint);
         }
 
         private void updateValues() => Scheduler.AddOnce(() =>
diff --git a/osu.Game/Overlays/Mods/ModFooterInformationDisplay.cs b/osu.Game/Overlays/Mods/ModFooterInformationDisplay.cs
index ca9c13e75f..b55717fc17 100644
--- a/osu.Game/Overlays/Mods/ModFooterInformationDisplay.cs
+++ b/osu.Game/Overlays/Mods/ModFooterInformationDisplay.cs
@@ -2,12 +2,10 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using osu.Framework.Allocation;
-using osu.Framework.Bindables;
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Colour;
 using osu.Framework.Graphics.Containers;
 using osu.Framework.Graphics.Shapes;
-using osu.Framework.Input.Events;
 using osu.Game.Graphics.UserInterface;
 using osuTK;
 using osuTK.Graphics;
@@ -20,8 +18,6 @@ public abstract partial class ModFooterInformationDisplay : CompositeDrawable
         protected FillFlowContainer RightContent { get; private set; } = null!;
         protected Container Content { get; private set; } = null!;
 
-        public BindableBool Collapsed { get; } = new BindableBool();
-
         private Container innerContent = null!;
 
         protected Box MainBackground { get; private set; } = null!;
@@ -30,8 +26,6 @@ public abstract partial class ModFooterInformationDisplay : CompositeDrawable
         [Resolved]
         private OverlayColourProvider colourProvider { get; set; } = null!;
 
-        private const float transition_duration = 250;
-
         [BackgroundDependencyLoader]
         private void load()
         {
@@ -88,7 +82,6 @@ private void load()
                             },
                             RightContent = new FillFlowContainer
                             {
-                                Alpha = 0,
                                 Origin = Anchor.CentreLeft,
                                 Anchor = Anchor.CentreLeft,
                                 AutoSizeAxes = Axes.X,
@@ -111,43 +104,6 @@ protected override void LoadComplete()
 
             Content.BorderColour = ColourInfo.GradientVertical(MainBackground.Colour, glowColour);
             innerContent.BorderColour = ColourInfo.GradientVertical(FrontBackground.Colour, glowColour);
-
-            Collapsed.BindValueChanged(_ =>
-            {
-                // Only start autosize animations on first collapse toggle. This avoids an ugly initial presentation.
-                startAnimating();
-                updateCollapsedState();
-            });
-
-            updateCollapsedState();
-        }
-
-        protected override bool OnHover(HoverEvent e)
-        {
-            startAnimating();
-            updateCollapsedState();
-            return true;
-        }
-
-        protected override void OnHoverLost(HoverLostEvent e)
-        {
-            updateCollapsedState();
-            base.OnHoverLost(e);
-        }
-
-        protected override bool OnMouseDown(MouseDownEvent e) => true;
-
-        protected override bool OnClick(ClickEvent e) => true;
-
-        private void startAnimating()
-        {
-            Content.AutoSizeEasing = Easing.OutQuint;
-            Content.AutoSizeDuration = transition_duration;
-        }
-
-        private void updateCollapsedState()
-        {
-            RightContent.FadeTo(Collapsed.Value && !IsHovered ? 0 : 1, transition_duration, Easing.OutQuint);
         }
     }
 }

```

peppy replied:
looks good

bdach committed:
Move collapsed bindable back to `BeatmapAttributesDisplay` for now

Better to have it only in one place that needs it, rather than have it
not work as expected when someone inherits
`ModFooterInformationDisplay`.
